dali/kernels/slice/slice_flip_normalize_permute_gpu.cuh

```diff
@@ -0,0 +1,273 @@
+#include "dali/core/cuda_error.h"
+#include "dali/core/dev_array.h"
+#include "dali/core/error_handling.h"
+#include "dali/kernels/common/copy.h"
+#include "dali/kernels/kernel.h"
+#include "dali/kernels/slice/slice_flip_normalize_permute_common.h"
+
+namespace dali {
+namespace kernels {
+
+namespace detail {
+
+template <size_t Dims>
+struct SampleDesc {
+  void *__restrict__ out;
+  const void *__restrict__ in;
+  DeviceArray<int64_t, Dims> in_strides;
+  DeviceArray<int64_t, Dims> out_strides;
+  DeviceArray<int64_t, Dims> out_shape;
+  DeviceArray<int64_t, Dims> padded_out_shape;
+};
+
+struct BlockDesc {
+  int sampleIdx;
+  size_t offset;
+  size_t size;
+};
+
+template <typename OutputType, typename InputType, unsigned Dims>
+__device__ inline void SliceFlipNormalizePermuteFunc(OutputType *__restrict__ out,
+                                                     const InputType *__restrict__ in,
+                                                     const int64_t *out_strides,
+                                                     const int64_t *in_strides,
+                                                     const int64_t *out_shape,
+                                                     const int64_t *padded_out_shape,
+                                                     bool should_normalize,
+                                                     bool should_zero_pad,
+                                                     unsigned norm_dim,
+                                                     const float *mean,
+                                                     const float *inv_stddev,
+                                                     size_t offset, size_t block_end) {
+  if (Dims > 1 && !should_normalize &&
+      out_strides[Dims - 1] == in_strides[Dims - 1] &&
+      out_shape[Dims - 1] == padded_out_shape[Dims - 1]) {
+    const unsigned NextDims = Dims > 1 ? Dims - 1 : 1;
+    SliceFlipNormalizePermuteFunc<OutputType, InputType, NextDims>(
+        out, in, out_strides, in_strides, out_shape, padded_out_shape, should_normalize,
+        should_zero_pad, norm_dim, mean, inv_stddev, offset, block_end);
+    return;
+  }
+
+  for (; offset < block_end; offset += blockDim.x) {
+    size_t idx = offset;
+    size_t out_idx = offset;
+    size_t in_idx = 0;
+    unsigned norm_i = 0;
+    bool zero_pad = false;
+
+    for (unsigned d = 0; d < Dims; d++) {
+      unsigned out_stride = static_cast<unsigned>(out_strides[d]);
+      unsigned i_d = idx / out_stride;
+      idx %= out_stride;
+      if (zero_pad = (should_zero_pad && i_d >= out_shape[d]))
+        break;
+
+      if (d == norm_dim)
+        norm_i = i_d;
+
+      in_idx += i_d * in_strides[d];
+    }
+
+    if (zero_pad) {
+      out[out_idx] = 0;
+    } else {
+      in_idx += idx;  // remaining dims have equal strides
+      if (should_normalize) {
+        out[out_idx] = clamp<OutputType>(
+          (static_cast<float>(in[in_idx]) - mean[norm_i]) * inv_stddev[norm_i]);
```

Comment:
Thing to try:
pass `norm_mean = -mean * inv_stddev` instead of `mean`. Then, `out = fmaf(in, inv_stddev, norm_mean)`

Reply:
done